### ISSUE TITLE
Click event not triggered from touchUp when waiting on script

### ISSUE THREAD

**Issue by kongbai1996** [servo/servo#34806]
**Describe the bug:**
When the click event is not triggered and there is a time-consuming operation in the touchstart listener

**To Reproduce:**
this is a test case:
```
<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
    <style>
        div {
            border: 2px black solid;
            width: 360px;
            height: 50px;
        }
    </style>
    <script>
        function fib(num) {
            if (num === 0) {
                return 0;
            }
            if (num === 1) {
                return 1;
            }
            return fib(num - 2) + fib(num - 1);
        }
    </script>
</head>

<body>
    <div id="div1" style="background-color: green;">div1 click here when fib</div>
    <div id="div2" style="background-color: green;">div2 click here when settimeout</div>
    <div id="result"></div>
</body>
<script>
    div1.addEventListener("touchstart", (e) => {
        fib(40);
        result.innerHTML += `<p>div1 touchstart fire</p>`
    })
    div1.addEventListener("click", (e) => {
        result.innerHTML += `<p>div1 click fire</p>`
    })
    div2.addEventListener("touchstart", (e) => {
        setTimeout(() => {
            result.innerHTML += `<p>div2 touchstart fire</p>`
        }, 1000)
    })
    div2.addEventListener("click", (e) => {
        result.innerHTML += `<p>div2 click fire</p>`
    })
</script>
</html>
```

**Platform:**
OpenHarmony

**Comment by kongbai1996** [comment]
Using chromium kernel
![image](https://github.com/user-attachments/assets/6655295c-1d33-441c-9922-bcec10031785)
Using servo kernel
![image](https://github.com/user-attachments/assets/478f6c78-c3ff-4220-9e9e-86690363b705)

**Comment by jschwe** [comment]
Did you test if it works as expected on Linux servo?

**Comment by kongbai1996** [comment]
> Did you test if it works as expected on Linux servo?
It's no problem on linux.

**Comment by kongbai1996** [comment]
On the oh, simulate_mouse_click is used to simulate a click event. On the Linux operating system, simulate_mouse_click is not used here. Maybe he'll have a similar problem on android.
https://github.com/servo/servo/blob/main/components/compositing/compositor.rs#L1596

**Comment by jschwe** [comment]
On the oh port we haven't installed the mouse event listeners yet. Should be simple to set up.

**Comment by kongbai1996** [comment]
When touchdown is triggered, the status of TouchHandler is set to WaitingForScript. If the callback function of touchstart takes a long time, WaitingForScript also lasts for a long time.
During this period, the touchup is triggered, but the status is still WaitingForScript. As a result, TouchAction::Click is not returned, and the click event cannot be triggered.
https://github.com/servo/servo/blob/main/components/compositing/touch.rs#L239

**Comment by kongbai1996** [comment]
![image](https://github.com/user-attachments/assets/746207dc-e478-448d-b39d-efcd721016a5)

**Comment by kongbai1996** [comment]
I want to split this branch in WaitingForScript and return TouchAction::Click, is that correct?
> https://github.com/servo/servo/blob/main/components/compositing/touch.rs#L239

**Comment by jschwe** [comment]
> During this period, the touchup is triggered, but the status is still WaitingForScript. As a result, TouchAction::Click is not returned, and the click event cannot be triggered.

That doesn't sound like an OpenHarmony specific issue - if you do something like `sleep(10)` in the script, then probably this should also affect servo on all platforms.

**Comment by kongbai1996** [comment]
> Maybe he'll have a similar problem on android. 
So I think there is a similar problem in Android, which I haven't tried.
However, this logic is not used in Linux. It directly triggers `MouseWindowEvent::Click` through `EmbedderEvent::MouseWindowEventClass`. The touchdown switchover status does not need to be identified, and it is not a click simulated by using `simulate_mouse_click`.

**Comment by jschwe** [comment]
Makes sense - I guess if we use a Linux/Mac/Windows system with a **touch screen**, then we would run into the same issue right?

**Comment by jschwe** [comment]
[The spec for touchend](https://w3c.github.io/touch-events/#event-touchend) says

> If, however, either the touchstart, touchmove or touchend event has been [canceled](https://w3c.github.io/touch-events/#dfn-canceled-event) during this interaction, **no mouse or click events will be fired**, and the resulting sequence of events would simply be:
> ....

If I understood this correctly, then the touchstart event handler can tell the UA to cancel the current touch action from the event handler, which would mean we need to wait for the script to finish before deciding to simulate a click event or not.
I'm not very sure I understood this correctly though, so I may very well be wrong.

**Comment by fabricedesre** [comment]
> Makes sense - I guess if we use a Linux/Mac/Windows system with a **touch screen**, then we would run into the same issue right?

I can test that on a pinephone - but that could take me a couple of days.

> If I understood this correctly, then the touchstart event handler can tell the UA to cancel the current touch action from the event handler, which would mean we need to wait for the script to finish before deciding to simulate a click event or not.
I'm not very sure I understood this correctly though, so I may very well be wrong.

That's also my understanding, and why I believe the proposed fix is not correct.

**Comment by kongbai1996** [comment]
> If I understood this correctly, then the touchstart event handler can tell the UA to cancel the current touch action from the event handler, which would mean we need to wait for the script to finish before deciding to simulate a click event or not. I'm not very sure I understood this correctly though, so I may very well be wrong.
Maybe you're talking about the preventDefault() method.
It is also not currently supported.
https://github.com/webbeef/servo/blob/2786f4e332f718dd57c1b4fffc60abf17d284138/components/compositing/touch.rs#L203

**Comment by kongbai1996** [comment]
> Makes sense - I guess if we use a Linux/Mac/Windows system with a **touch screen**, then we would run into the same issue right?

Maybe, I didn't try

**Comment by gterzian** [comment]
> If the callback function of touchstart takes a long time, WaitingForScript also lasts for a long time

What I find interesting is that in your example the callback does not take a long time, rather it triggers yet another callback via the timeout.

So we're not talking about blocking script with something that takes a long time, rather we're scheduling an async timeout that will call into the callback after 1s. 

How come we're still waiting for script after `touchstart` was fired, but the timeout hasn't yet?

**Comment by kongbai1996** [comment]
> How come we're still waiting for script after touchstart was fired, but the timeout hasn't yet?

The timeout example does not block state transitions because it is asynchronous.
> EDIT: I now see the fib(40); call in the other touchstart handler... So that's the blocking call.

However, in fib(40), the state is kept in `WaitingForScript`.
